### DISCUSSION TITLE
Poetry and Python not available in `postUpgradeTasks`

### DISCUSSION THREAD

**Discussion by CCardosoDev** [renovatebot/renovate#28982] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub with version 37.183.1

### Please tell us more about your question or problem

I am trying to update the Changelog using `postUpgradeTasks` for every new Renovate PR. For that, I want to be able to execute a bash or Python script, since the version calculation and positioning has some complexity, but I am encountering some issues:
- executing a Python script using Poetry - any Python or Poetry command I am getting the error
```
Command failed: poetry --version
/bin/sh: 1: poetry: not found
```
or 
```
Command failed: python --version
/bin/sh: 1: python: not found
```
- using a bash script - I don't want to have the same script in every repo so I need to be able to download it from somewhere

My package rule is:
```
{
      "postUpgradeTasks": {
        "commands": [
          "echo \"test\" >> CHANGELOG.md",
          "poetry --version"
        ],
        "executionMode": "branch"
      },
      "matchManagers": ["poetry"]
}
```

We use the slim Docker image and `binarySource` is set to the default `install`.

I don't fully understand how containerbase works, but from my understanding `Poetry` should be installed and available at runtime.

Do you have any hints on what could be the issue here? Or do you think this can be achieved in any other way?

**Comment by rarkins** [comment]
If you run your script only after Renovate updates lock files, it should already be installed. If not, run `install-tool poetry x.y.z` yourself as part of your script. Alternatively, use the full image which has every manager pre-installed.

**Reply by CCardosoDev** [reply, marked as answer]
This was a great hint, now I understand why I don't have Poetry or Python installed, because the repo does not have a lock file since this project is a Python package and does not need one, only for testing. This means that instructions like

```
DEBUG: Executing command (repository=Bot/renovate-testing.7IYlBR, branch=renovate/my-package)
       "command": "install-tool poetry 1.8.2"
```
Are not executed during the upgrade.

This also means that I need to install it explicitly both `Python` and `Poetry` with something like:

```
{
      "postUpgradeTasks": {
        "commands": [
          "install-tool python 3.11.9",
          "install-tool poetry 1.8.2",
          "echo \"test\" >> CHANGELOG.md",
          "poetry --version"
        ],
        "executionMode": "branch"
      },
      "matchManagers": ["poetry"]
}
```

This works as expected, however, I would have to add some logic to understand which version of Python to use.

Thanks a lot for the quick response, this already helps me a lot.

**Comment by HonkingGoose** [comment]
> ### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.
> 
> GitHub with version 37.183.1

By the way, you should update. Your version was released on the 12th of February. That's nearly three months old.

**Reply by CCardosoDev** [reply]
Yes, we are in the process of doing so.
